src/Workspaces/Core/Desktop/Workspace/MSBuild/ProjectFile/ProjectFile.cs

```diff
@@ -185,13 +186,35 @@ private static Task<MSB.Execution.BuildResult> BuildAsync(MSB.Execution.BuildMan
                     {
                         buildManager.CancelAllSubmissions();
                         buildManager.EndBuild();
-                        registration.Dispose();
                     }
                     finally
                     {
                         taskSource.TrySetCanceled();
                     }
                 });
+
+                // Dispose of the registration as soon as we no longer need it
+                var _ = DisposeWhenCompleteAsync(taskSource.Task, registration);
+
+                async Task DisposeWhenCompleteAsync<T>(Task task, T disposable)
```

Comment:
I wouldn't mind this as a general extension on Task that goes with our other task extensions.  It seems out of place here to be generic, but only called with one arg in one place.